pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -35,28 +38,42 @@ type ClusterMemberController struct {
 	podLister      corev1listers.PodLister
 	nodeLister     corev1listers.NodeLister
 	networkLister  configv1listers.NetworkLister
+	// machineAPIChecker determines if the precondition for this controller is met,
+	// this controller can be run only on a cluster that exposes a functional Machine API
+	machineAPIChecker ceohelpers.MachineAPIChecker
+
+	masterMachineLister   machinelistersv1beta1.MachineLister
+	masterMachineSelector labels.Selector
 }
 
 func NewClusterMemberController(
 	operatorClient v1helpers.OperatorClient,
+	machineAPIChecker ceohelpers.MachineAPIChecker,
 	kubeInformers v1helpers.KubeInformersForNamespaces,
 	networkInformer configv1informers.NetworkInformer,
+	masterMachineInformer cache.SharedIndexInformer,
+	masterMachineSelector labels.Selector,
 	etcdClient etcdcli.EtcdClient,
 	eventRecorder events.Recorder,
 ) factory.Controller {
 	c := &ClusterMemberController{
-		operatorClient: operatorClient,
-		etcdClient:     etcdClient,
-		podLister:      kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Pods().Lister(),
-		nodeLister:     kubeInformers.InformersFor("").Core().V1().Nodes().Lister(),
-		networkLister:  networkInformer.Lister(),
+		operatorClient:        operatorClient,
+		machineAPIChecker:     machineAPIChecker,
+		etcdClient:            etcdClient,
+		podLister:             kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Pods().Lister(),
+		nodeLister:            kubeInformers.InformersFor("").Core().V1().Nodes().Lister(),
+		networkLister:         networkInformer.Lister(),
+		masterMachineLister:   machinelistersv1beta1.NewMachineLister(masterMachineInformer.GetIndexer()),
+		masterMachineSelector: masterMachineSelector,
 	}
-	return factory.New().ResyncEvery(time.Minute).WithInformers(
-		kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Pods().Informer(),
-		kubeInformers.InformersFor("").Core().V1().Nodes().Informer(),
-		networkInformer.Informer(),
-		operatorClient.Informer(),
-	).WithSync(c.sync).ResyncEvery(time.Minute).ToController("ClusterMemberController", eventRecorder.WithComponentSuffix("cluster-member-controller"))
+	return factory.New().ResyncEvery(time.Minute).
+		WithBareInformers(networkInformer.Informer()).
+		WithInformers(
+			kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Pods().Informer(),
+			kubeInformers.InformersFor("").Core().V1().Nodes().Informer(),
+			operatorClient.Informer(),
+			masterMachineInformer,
+		).WithSync(c.sync).WithSyncDegradedOnError(operatorClient).ResyncEvery(time.Minute).ToController("ClusterMemberController", eventRecorder.WithComponentSuffix("cluster-member-controller"))
 }
 
 func (c *ClusterMemberController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
```

Comment:
don't set the condition in this message, just return the error from reconcileMembers

Reply:
oh, right

Reply:
I think that we don't have to clean up the old condition because the new ones will have the same name

Reply:
yeah,
```
// WithSyncDegradedOnError encapsulate the controller sync() function, so when this function return an error, the operator client
// is used to set the degraded condition to (eg. "ControllerFooDegraded"). The degraded condition name is set based on the controller name.
```